float64[] control
int32 type
